Add github_branch data and resource

cornfeedhobo opened the pull request:
This PR adds a new resource for creating and managing Github repository branches. I find this useful for modules that codify Gitflow.

Please let me know if I can fix anything.

```bash
$ TF_ACC=1 go test github.com/terraform-providers/terraform-provider-github/github \
    -v -timeout 120m -run=TestAccGithubBranchDataSource_basic
=== RUN   TestAccGithubBranchDataSource_basic
=== PAUSE TestAccGithubBranchDataSource_basic
=== CONT  TestAccGithubBranchDataSource_basic
--- PASS: TestAccGithubBranchDataSource_basic (9.28s)
PASS
ok      github.com/terraform-providers/terraform-provider-github/github (cached)

$ TF_ACC=1 go test github.com/terraform-providers/terraform-provider-github/github \
    -v -timeout 120m -run=TestAccGithubBranch_basic
=== RUN   TestAccGithubBranch_basic
=== PAUSE TestAccGithubBranch_basic
=== CONT  TestAccGithubBranch_basic
--- PASS: TestAccGithubBranch_basic (10.20s)
PASS
ok      github.com/terraform-providers/terraform-provider-github/github 10.210s

$ TF_ACC=1 go test github.com/terraform-providers/terraform-provider-github/github \
    -v -timeout 120m -run=TestAccGithubBranch_withSourceBranch
=== RUN   TestAccGithubBranch_withSourceBranch
=== PAUSE TestAccGithubBranch_withSourceBranch
=== CONT  TestAccGithubBranch_withSourceBranch
--- PASS: TestAccGithubBranch_withSourceBranch (6.91s)
PASS
ok      github.com/terraform-providers/terraform-provider-github/github 6.919s
```

cornfeedhobo commented:
Partially addresses #286

shoekstra commented on github/data_source_github_repository_branch_test.go:
```diff
@@ -0,0 +1,46 @@
+  repository = github_repository.%s.name
+}
+`, name, repo, name, name)
+}
```
You only need to specify the `github_repository_branch` data resource here as a test repo is expected to already exist, so this can be simplified by removing the `github_repository` and using `repository = "test-repo"` in the `github_repository_branch` resource.

cornfeedhobo replied:
same comment as above.

cornfeedhobo commented on github/data_source_github_repository_branch.go:
```diff
@@ -0,0 +1,63 @@
+				ForceNew: true,
+				Default:  "master",
+			},
+
```
Just pulling an example at random, there are other resources that include a blank line between computed and non-computed attributes. e.g. [github_repository](https://github.com/terraform-providers/terraform-provider-github/blob/c533ed9cbc2c17887d150cafcea0509bfcd78b08/github/resource_github_repository.go#L112).

shoekstra replied:
Yea i see that now.. it would be nice if there was some standard because it feels random at the moment (e.g. why separate computed fields but not required or optional).

Please ignore this comment and any others relating to removing lines between attribute names then 👍

cornfeedhobo commented on github/resource_github_repository_branch.go:
```diff
@@ -0,0 +1,167 @@
+				ForceNew: true,
+				Computed: true,
+			},
+
```
Same comment as above.

cornfeedhobo commented on github/resource_github_repository_branch_test.go:
```diff
@@ -0,0 +1,138 @@
+  branch     = "%s"
+}
+`, name, repo, name, name, branch)
+}
```
How come other tests do this then?

shoekstra replied:
[In the README under Test repositories](https://github.com/terraform-providers/terraform-provider-github#test-repositories) says "In the organization you are using above, create the following test repositories", so there is the expectation they should exist when writing tests.

I can appreciate other tests are creating a repository, and they should also corrected, but there is no reason IMO to merge new contributions that will later need to be fixed.

cornfeedhobo replied:
It looks like indeed the fixtures aren't touched. I've updated the tests accordingly.

cornfeedhobo commented:
@shoekstra Thanks for the review. I left some questions for you to clarify, but in general I'm happy to change anything you want. All your points are minor, including the resource name. github_branch is fine with me. Let me know what you think or if you want to wait for @jcudit before I make any follow up changes.

Regards

cornfeedhobo commented:
@shoekstra Alright. I renamed the new resources and addressed your points about testing. Have another look and let me know if I can polish anything else up. Thanks again!

cornfeedhobo commented:
@shoekstra Alright. I renamed the new resources and addressed your points about testing. However, it is unclear to me why the tests are failing. Any ideas?

cornfeedhobo commented:
@shoekstra Could you or someone else please look at this? The build passes locally.

cornfeedhobo commented:
@shoekstra @jcudit 

Bump. I addressed all the concerns I think. :pray:

jcudit commented:
👋 @cornfeedhobo - this is looking good so far.  I think all that is missing is to update the `website/` directory with changes similar to these two previous examples:
- https://github.com/terraform-providers/terraform-provider-github/pull/362/files#diff-ad5e3f4597f97730fd542a7e8e0678f8
- https://github.com/terraform-providers/terraform-provider-github/pull/362/files#diff-876606ce80d645e2afa28ba4a9bf7b2f

Docs for both the datasource and resource would be helpful.

Also, I am in favour of `github_repository_branch` because it feels more organized and less ambiguous.

shoekstra commented:
Sincere apologies for the delayed response! I haven't had time to look at the test errors, I'll see if I can do that later today... 

> Also, I am in favour of `github_repository_branch` because it feels more organized and less ambiguous.

I worry this leads to bloated resource names and we should only include `repository` in the resource name if there is also a non repository version of the resource, and we already have `github_branch_protection`, so `github_branch` feels a better fit than `github_repository_branch`.

But I don't want to let a naming discussion block this functionality, so I'll create a separate issue to discuss. 🙂

cornfeedhobo commented:
I'm going to keep going with `github_branch`, but I do encourage a separate issue being opened about this, because while github_repository_branch*` is verbose, I'm pretty sure that's a pill everyone has swallowed in the context of Terraform.

cornfeedhobo commented:
I'm going to verify importing before signing off on this. stand by.

shoekstra commented:
> I'm going to verify importing before signing off on this. stand by.

You can add a test for the importer functionality in your resource's integration tests, [here's an example](https://github.com/terraform-providers/terraform-provider-github/blob/master/github/resource_github_repository_file_test.go#L142-L146).

cornfeedhobo commented:
@shoekstra interesting. It looks like my code already does that, but then doing some functional tests and I see that after import, the subsequent apply is attempting to change _waayy_ too many attributes. Is it possible the test suites should have a follow up `apply` step to check for drift after import?

shoekstra commented:
This is generally solved by adding those attributes to the resource's Read func so when the resource is refreshed those values are read and then stored in the state.

You'll notice in `resourceGithubRepositoryFileCreate` for example that instead of setting values in the state file it calls [`resourceGithubRepositoryFileRead`](https://github.com/terraform-providers/terraform-provider-github/blob/master/github/resource_github_repository_file.go#L172) which then reads the updated resource and sets it's values in the state file. AFAIK you want to set as much as you can via the Read func so that changes are captured when refreshing or importing an existing resource.

Hope that helps!

cornfeedhobo commented:
Fixes #401

cornfeedhobo commented:
@jcudit @shoekstra thanks for hanging in there! I believe this is ready now.

cornfeedhobo commented on github/resource_github_branch_test.go:
```diff
@@ -0,0 +1,127 @@
+			},
+			{
+				ResourceName:      rn,
+				ImportState:       true,
```
I was wondering if that was the case.

cornfeedhobo commented on github/resource_github_branch.go:
```diff
@@ -0,0 +1,181 @@
+				if err != nil {
+					return nil, err
+				}
+				sourceBranch := "master"
```
I was thinking about this too, but honestly thought I'd hear opposition! :p
Let me take a swing at it right now.

cornfeedhobo commented:
@shoekstra Thanks for the review. I think I've addressed everything you've brought up. It also gave me the chance to clean up some other minor things. Have another look please, and I'll try to be fast about any final changes. Thanks again!

shoekstra commented on github/resource_github_branch.go:
```diff
@@ -0,0 +1,202 @@
+		if err != nil {
+			return nil, err
+		}
+		d.SetId(buildTwoPartID(&repoName, &branchName))
```
This line can be removed right? The resource ID won't change, you're only checking if the `source_branch` field should be set.

shoekstra commented on github/resource_github_branch.go:
```diff
@@ -0,0 +1,202 @@
+}
+
+func resourceGithubBranchImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	repoName, branchName, err := parseTwoPartID(d.Id(), "repository", "branch")
```
```suggestion
	_, branchName, err := parseTwoPartID(d.Id(), "repository", "branch")
```

`repoName` is not needed as you don't need to set the ID [here](https://github.com/terraform-providers/terraform-provider-github/pull/364/files#diff-81bb68b8177b7f989f656af57eebc7afR196).

shoekstra commented on github/resource_github_branch_test.go:
```diff
@@ -0,0 +1,204 @@
+				ImportState:       true,
+				ImportStateId:     fmt.Sprintf("%s:%s", repo, branch),
+				ImportStateVerify: true,
+				ImportStateVerifyIgnore: []string{
```
The OCD in me says the previous for 4 values should be in line with the beginning of this one.

cornfeedhobo replied:
Same, but you'll have to take that one up with `fmt`. This is out of my hands. Even your linter says so.

shoekstra replied:
😆 haha fine!

shoekstra commented on website/docs/r/branch.html.markdown:
````diff
@@ -0,0 +1,54 @@
+
+```
+$ terraform import github_branch.terraform terraform:master
+```
````
You should include an example about the optional source branch, something like:

> The source branch can be specified if it was not `master`, e.g.
> 
> ```
> $ terraform import github_branch.terraform terraform:test:some_branch
> ```

I'll leave the wording up to you.

cornfeedhobo commented:
@shoekstra updated

jcudit commented:
👋 I've got this one on my radar and am happy with the outcome so far.  Great work 😄 

I'm focusing on merging a client change that would cause a minor conflict here when merged.  Afterwards, I'll give this branch a thorough review pass and will either get it merged with the upcoming release or the one that follows.  Thanks again for the contribution and patience.

cornfeedhobo commented:
Thanks! We have some work that is probably 2 weeks out, and it would be great to have this in by then.

jcudit commented on github/data_source_github_branch.go:
```diff
@@ -0,0 +1,69 @@
+			orgName, repoName, branchRefName, err)
+	}
+
+	d.SetId(buildTwoPartID(&repoName, &branchName))
```
```suggestion
	d.SetId(buildTwoPartID(repoName, branchName))
```

jcudit commented on github/data_source_github_branch_test.go:
```diff
@@ -0,0 +1,74 @@
+					resource.TestCheckResourceAttrSet(rn, "sha"),
+				),
+			},
+			{
+				Config: testAccCheckGithubBranchDataSourceConfig(name, repo, "master"),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(rn, "repository", repo),
+					resource.TestCheckResourceAttr(rn, "branch", "master"),
+					resource.TestCheckResourceAttrSet(rn, "etag"),
+					resource.TestCheckResourceAttrSet(rn, "ref"),
+					resource.TestCheckResourceAttrSet(rn, "sha"),
+				),
+			},
```
Is this a duplicate of the block above?

cornfeedhobo replied:
@jcudit Yes. My understanding is that this checks for drift, essentially making sure my state doesn't accidentally drift between runs. This helped me catch an error with where I was setting `etag`.

jcudit commented on github/data_source_github_branch_test.go:
```diff
@@ -0,0 +1,74 @@
+				),
+			},
+			{
+				Config: testAccCheckGithubBranchDataSourceConfig(name, repo, "test-branch"),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(rn, "repository", repo),
+					resource.TestCheckResourceAttr(rn, "branch", "test-branch"),
+					resource.TestCheckResourceAttrSet(rn, "etag"),
+					resource.TestCheckResourceAttrSet(rn, "ref"),
+					resource.TestCheckResourceAttrSet(rn, "sha"),
+				),
+			},
```
Is this a duplicate of the block above?

cornfeedhobo replied:
same answer as above.

jcudit commented on github/provider.go:
```diff
@@ -69,6 +70,8 @@ func Provider() terraform.ResourceProvider {
 		},
 
 		DataSourcesMap: map[string]*schema.Resource{
+			"github_actions_public_key": dataSourceGithubActionsPublicKey(),
```
Thanks for organizing this.

cornfeedhobo replied:
no problem!

jcudit commented on github/resource_github_branch.go:
```diff
@@ -0,0 +1,201 @@
+		ctx = context.WithValue(ctx, ctxId, d.Id())
+	}
+
+	client := meta.(*Organization).client
```
```suggestion
	client := meta.(*Organization).v3client
```

jcudit commented on github/resource_github_branch.go:
```diff
@@ -0,0 +1,201 @@
+			orgName, repoName, branchRefName, err)
+	}
+
+	d.SetId(buildTwoPartID(&repoName, &branchName))
```
```suggestion
	d.SetId(buildTwoPartID(repoName, branchName))
```

jcudit commented on github/resource_github_branch.go:
```diff
@@ -0,0 +1,201 @@
+		ctx = context.WithValue(ctx, ctxEtag, d.Get("etag").(string))
+	}
+
+	client := meta.(*Organization).client
```
```suggestion
	client := meta.(*Organization).v3client
```

jcudit commented on github/resource_github_branch_test.go:
```diff
@@ -0,0 +1,204 @@
+				ImportStateId:     fmt.Sprintf("%s:%s", repo, branch),
+				ImportStateVerify: true,
+				ImportStateVerifyIgnore: []string{
+					"source_sha",
```
What does work around?

cornfeedhobo replied:
`source_sha` is extremely difficult to determine at import time. I dug through github's api and git. It's non-trivial to find the parent commit, and even once you do, it's not guaranteed accurate. It my testing, I was getting 
```
testing.go:654: Step 2 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
        
        (map[string]string) {
        }
        
        
        (map[string]string) (len=1) {
         (string) (len=10) "source_sha": (string) (len=40) "18edb0e04930c3e4cff23fb837d3116898e0c6a6"
        }
```

jcudit commented on github/resource_github_branch_test.go:
```diff
@@ -0,0 +1,204 @@
+			return fmt.Errorf("Expected ID to be %v, got %v", id, rs.Primary.ID)
+		}
+
+		conn := testAccProvider.Meta().(*Organization).client
```
```suggestion
		conn := testAccProvider.Meta().(*Organization).v3client
```

jcudit commented on website/docs/d/branch.html.markdown:
````diff
@@ -0,0 +1,37 @@
+```hcl
+data "github_branch" "master" {
+  repository = "example"
+  branch     = "development"
````
```suggestion
  branch     = "master"
```

Could also align the data resource name to `development`.

jcudit commented on github/resource_github_branch_test.go:
```diff
@@ -0,0 +1,204 @@
+			continue
+		}
+
+		conn := testAccProvider.Meta().(*Organization).client
```
```suggestion
		conn := testAccProvider.Meta().(*Organization).v3client
```

cornfeedhobo commented:
@jcudit The acceptance tests already ran here and locally - everything was fine. Did something change in master necessitating these? What's up - I'm lost.

jcudit commented:
Sorry for the confusion @cornfeedhobo, there were changes merged around client and pointer use. These are the minor conflicts [mentioned above](https://github.com/terraform-providers/terraform-provider-github/pull/364#issuecomment-613703472).

Not sure about the failing test as yet.

cornfeedhobo commented:
Gotchya. On it!

cornfeedhobo commented:
@jcudit I've addressed most everything and left a comment on the import state question.

Unfortunately it look like your travis config is broken right now. Ping me with your review and I'll kick off travis with an empty commit if all is good. Thanks!

jcudit commented:
@cornfeedhobo - got it sorted, apologies for that interrupt.  Can you try a rebase so we can validate tests are passing?

cornfeedhobo commented:
@jcudit Look at that happy travis! thanks!

jcudit commented:
Happy to see this one ship!  Great feature, great collaboration and a lot of patience to get to this point.  Thanks again for your efforts.

jacobisaliveandwell commented:
Likewise, thanks a bunch!

cornfeedhobo commented:
Likewise. Thanks for the thoughtful reviews!
